### PR TITLE
enhance(main/libusb): Add integration with `termux-usb -E`

### PR DESCRIPTION
**Pull request by kyufie**
This commit adds the ability for `libusb` to access connected USB devices through file descriptor given by `termux-usb -E`, extending its usability for non-root users.

In the presence of `TERMUX_USB_FD`, `libusb` will try to read the fd number given and add it to the list of available devices, disregarding other discovery methods.
Otherwise it will try to use the usual `/dev/bus/usb`-based device discovery so that root users can still enjoy the benefits of having root access.

To test it, run `termux-usb -r -E -e 'command' device_name`. Make sure that `command` uses `libusb`.

### CODE REVIEW

**Comment by kyufie** [comment]
Here's a little program that I made to help testing this library, compile with `-lusb-1.0`.
All that is needed is a USB drive.
Once it's running with the drive connected, the kernel driver should be detached from the drive and the drive should disappear from the operating system.
```c
#include <libusb-1.0/libusb.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
    int status;
    size_t devnum;
    struct libusb_device **devlist;
    struct libusb_device_handle *devhdl;

    libusb_set_debug(NULL, 3);

    status = libusb_init(NULL);
    if (status < 0) {
        puts(libusb_error_name(status));
        return 1;
    }

    devnum = libusb_get_device_list(NULL, &devlist);
    printf("Number of device attached: %lu\n", devnum);
    if (devnum == 0)
        return 1;

    status = libusb_open(devlist[0], &devhdl);
    if (status < 0) {
        puts(libusb_error_name(status));
        return 1;
    }

    status = libusb_detach_kernel_driver(devhdl, 0);
    if (status < 0) {
        puts(libusb_error_name(status));
        return 1;
    }

    return EXIT_SUCCESS;
}
```

**Comment by hansm629** [comment]
@kyufie 
Hello sir, thank you for your hard work in development.
If I understand correctly, does this mean that now in the `Termux environment`, `USB storage` can be recognized without needing root access?

**Comment by twaik** [comment]
No. Termux does not have any special rights on system, it is a regular Android app and it can not access HID or Mass Storage devices using USB Host API.

**Comment by kyufie** [comment]
@hansm629
As of now, it's only useful for accessing other device through ADB.

**Comment by IntinteDAO** [comment]
> @kyufie Hello sir, thank you for your hard work in development. If I understand correctly, does this mean that now in the `Termux environment`, `USB storage` can be recognized without needing root access?

But it seems to me that you can access files on a USB device without root, simply by symlinking the directory mounted to termux.

Not a good option, but a workaround if you mean it

**Comment by hansm629** [comment]
@IntinteDAO 
If that's possible, I believe it would be a major turning point for computing in Termux.
The only thing I found lacking in the Termux computing environment was the inability to mount USB storage.

I really hope it becomes possible.

**Comment by twaik** [comment]
> I really hope it becomes possible.

It will not be possible without root.

**Comment by IntinteDAO** [comment]
> > I really hope it becomes possible.
> 
> It will not be possible without root.

> @IntinteDAO If that's possible, I believe it would be a major turning point for computing in Termux. The only thing I found lacking in the Termux computing environment was the inability to mount USB storage.
> 
> I really hope it becomes possible.

I was sure it could be done from within Termux directly, however, I actually have a problem with that, so I suspect I may have had root.

On the other hand, other apps can access the USB flash drive (e.g. Ghost Commander), so it's probably a technically doable issue.

**Comment by twaik** [comment]
> On the other hand, other apps can access the USB flash drive (e.g. Ghost Commander), so it's probably a technically doable issue.

They do not use usb api for that. They use storage API.

**Comment by IntinteDAO** [comment]
> > On the other hand, other apps can access the USB flash drive (e.g. Ghost Commander), so it's probably a technically doable issue.
> 
> They do not use usb api for that. They use storage API.

I assume so. Is there an option for Termux to use Storage API to read data over USB or not? From what I understand, no (no one has implemented it).

**Comment by kyufie** [comment]
> > > On the other hand, other apps can access the USB flash drive (e.g. Ghost Commander), so it's probably a technically doable issue.
> > 
> > 
> > They do not use usb api for that. They use storage API.
> 
> I assume so. Is there an option for Termux to use Storage API to read data over USB or not? From what I understand, no (no one has implemented it).

I usually just read the available mount points and figure out the path to the external storage from it.
Then, I use that path to access the external storage directly.
As long as you have access to that path, this shouldn't be a problem. Works for me to say the least.
`mount | awk '/\/mnt\/media_rw/ && split($3, a, /\//){print "/storage/" a[4]}'`

**Comment by IntinteDAO** [comment]
> > > > On the other hand, other apps can access the USB flash drive (e.g. Ghost Commander), so it's probably a technically doable issue.
> > > 
> > > 
> > > They do not use usb api for that. They use storage API.
> > 
> > 
> > I assume so. Is there an option for Termux to use Storage API to read data over USB or not? From what I understand, no (no one has implemented it).
> 
> I usually just read the available mount points and figure out the path to the external storage from it. Then, I use that path to access the external storage directly. As long as you have access to that path, this shouldn't be a problem. Works for me to say the least. `mount | awk '/\/mnt\/media_rw/ && split($3, a, /\//){print "/storage/" a[4]}'`

For external storage there is support in Termux thanks to termux-setup-storage. Termux, on the other hand, does not have access to USB storage (without root), although it is theoretically possible for it to have (someone would have to implement some “driver” (?) fuse / gvfs.

**Comment by agnostic-apollo** [comment]
https://github.com/termux/termux-app/issues/71#issuecomment-1869222653

**Comment by kyufie** [comment]
Qemu USB passthrough is now possible thanks to this PR.
I managed to connect my USB flash drive to my Windows VM and access it from there.
Screenshot_2024-10-14-23-24-01-534_com realvnc viewer android
Screenshot_2024-10-14-23-31-56-178_com realvnc viewer android

`termux-usb -r -E -e 'qemu-system-x86_64 -hda /sdcard/win7.qcow2 -m 4096 -accel tcg,thread=multi -smp 2 -device usb-ehci -device usb-host,hostbus=1,hostaddr=2 -vnc :0' /dev/bus/usb/001/002`
Adjust `hostbus` and `hostaddr` according to the numbers present in `/dev/bus/usb/...`

In theory, this should permit unbounded access to USB devices without root access.
> although it is theoretically possible for it to have (someone would have to implement some “driver” (?) fuse / gvfs.

This is probably the "driver" that you need.

I'm also thinking about other possibilities this might opens up (obtain root access without PC, arduino, all those things that people have been struggling with).

Someone seems to have already done it, but it was quite janky (#19635).

**Comment by hansm629** [comment]
@kyufie 
It seems to be working! That's great news.

Ultimately, I hope that USB storage can be freely mounted in both the Termux XFCE4 desktop environment and PRoot Linux!

**Comment by kyufie** [comment]
> > > > > On the other hand, other apps can access the USB flash drive (e.g. Ghost Commander), so it's probably a technically doable issue.
> > > > 
> > > > 
> > > > They do not use usb api for that. They use storage API.
> > > 
> > > 
> > > I assume so. Is there an option for Termux to use Storage API to read data over USB or not? From what I understand, no (no one has implemented it).
> > 
> > 
> > I usually just read the available mount points and figure out the path to the external storage from it. Then, I use that path to access the external storage directly. As long as you have access to that path, this shouldn't be a problem. Works for me to say the least. `mount | awk '/\/mnt\/media_rw/ && split($3, a, /\//){print "/storage/" a[4]}'`
> 
> For external storage there is support in Termux thanks to termux-setup-storage. Termux, on the other hand, does not have access to USB storage (without root), although it is theoretically possible for it to have (someone would have to implement some “driver” (?) fuse / gvfs.

If you're talking about access to the *files* inside the USB storage, as opposed to raw access, that one-liner I gave also works for USB drive.
USB drive counts as external storage too.

**Comment by hansm629** [comment]
@kyufie 

- libusb_1.0.27-1_aarch64.deb
- libusb-static_1.0.27-1_aarch64.deb

After installing these two packages, which are currently in a pull request, do I just need to add the following environment variables and start XFCE4?

```
export mount | awk '/\/mnt\/media_rw/ && split($3, a, /\//){print "/storage/" a[4]}'
```

**Comment by kyufie** [comment]
@hansm629
No that's not for envvar.
It's a command that gives you a list of path to the directory where your files in the external storage are located. You can `cd` into it, and do something just like you would on a regular directory.
```
$ mount | awk '/\/mnt\/media_rw/ && split($3, a, /\//){print "/storage/" a[4]}'
/storage/9016-4EF8 <-- External storage 1
/storage/ABCD-EFG <-- External storage 2
$ ls /storage/9016-4EF8
<Shows the contents of the external storage>
```
In the desktop environment, you should be able to type in the path in the file manager to go straight to that directory.

`libusb` is not required in this case.

**Comment by hansm629** [comment]
@kyufie
It is displaying as shown below. Is there something I am missing?

```
$ mount | awk '/\/mnt\/media_rw/ && split($3, a, /\//){print "/storage/" a[4]}'
/storage/7EEB-CB06
$ ls /storage/7EEB-CB06
"/storage/7EEB-CB06": No such file or directory (os error 2)
```

**Comment by kyufie** [comment]
@hansm629 
Perhaps the drive got disconnected by the time you use `ls`.
Also, if there's a prompt raised by an app asking to use the drive, reject it, that app may claim and detach the drive from the operating system.

**Comment by hansm629** [comment]
@kyufie 
I still cannot access the storage. Are there any additional steps I should take?

The steps I followed are as follows:

- Run **Termux**
- Installed **libusb_1.0.27-1**/**ibusb-static_1.0.27-1**
- Inserted the **USB storage** into the USB OTG
- Ran the following command
```
~ $ mount | awk '/\/mnt\/media_rw/ && split($3, a, /\//){print "/storage/" a[4]}'
/storage/7EEB-CB06
~ $ ls /storage/7EEB-CB06
ls: cannot access '/storage/7EEB-CB06': No such file or directory
``` 

Is there anything wrong or missing from the steps I took?

Could the USB storage format also have an effect?

**Comment by kyufie** [comment]
@hansm629
Not sure about that.
Can you do `mount | grep 7EEB-CB06` with the drive connected?
And what Android version are you using?

**Comment by kyufie** [comment]
Also:
>WRITE_EXTERNAL_STORAGE vs MANAGE_EXTERNAL_STORAGE
Enabling Android Settings -> Apps -> Termux -> Permissions -> Files will only grant legacy WRITE_EXTERNAL_STROAGE permission and enabling Android Settings -> Apps -> Special app access -> All files access -> Termux will only grant MANAGE_EXTERNAL_STORAGE permission. They are two different permissions, but access to primary external storage permission should still be available with either under the /storage/emulated/<user_id> or /sdcard paths.

>However, granting MANAGE_EXTERNAL_STORAGE should additionally grant access to unreliable/removable volumes like USB OTG devices under the /mnt/media_rw/XXXX-XXXX paths on Android >= 12, check below. The termux-setup-storage command only requests the WRITE_EXTERNAL_STROAGE if the Termux app is using targetSdkVersion < 30, which is 28 currently for F-Droid and GitHub releases.

https://github.com/termux/termux-app/issues/71#issuecomment-1869222653

**Comment by hansm629** [comment]
@kyufie 
Thank you so much for providing such a detailed response despite being busy. Should I try installing version `0.118.1` from the Termux-app GitHub release and test it again? I’ll update you after I’ve done the update and tested it! :)

**Comment by agnostic-apollo** [comment]
You need https://github.com/termux/termux-app/releases/tag/v0.119.0-beta.1 from github/fdroid for the `MANAGE_EXTERNAL_STORAGE` permission.

**Comment by hansm629** [comment]
@agnostic-apollo 
Hello Sir,
I tried again after installing `Termux v0.119.0-beta.1`, but the same issue persists. 

I activated the **MANAGE_EXTERNAL_STORAGE** permission and manually disabled the **WRITE_EXTERNAL_STORAGE** permission, but is there something else I might be missing?

Could it be that it's not working because it's without root?

Here is the termux-info:

Screenshot_20241018_002900_Settings
Screenshot_20241018_002929_Permission controller

```
$ mount | awk '/\/mnt\/media_rw/ && split($3, a, /\//){print "/storage/" a[4]}'
/storage/7EEB-CB06
$ ls /storage/7EEB-CB06
"/storage/7EEB-CB06": No such file or directory (os error 2)
``` 
```
Termux Variables:
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP__APK_FILE=/data/app/~~xN6e9m3kSw38MFPf1yepwg==/com.termux-x6aA1bFU2_ANGtnHxSWsqg==/base.apk
TERMUX_APP__APK_RELEASE=GITHUB
TERMUX_APP__APP_VERSION_CODE=1020
TERMUX_APP__APP_VERSION_NAME=0.119.0-beta.1
TERMUX_APP__DATA_DIR=/data/user/0/com.termux
TERMUX_APP__IS_DEBUGGABLE_BUILD=true
TERMUX_APP__IS_INSTALLED_ON_EXTERNAL_STORAGE=false
TERMUX_APP__PACKAGE_NAME=com.termux
TERMUX_APP__PID=27696
TERMUX_APP__TARGET_SDK=28
TERMUX_VERSION=0.119.0-beta.1
TERMUX__SE_FILE_CONTEXT=u:object_r:app_data_file:s0:c127,c257,c512,c768
TERMUX__SE_INFO=default:targetSdkVersion=28:complete
TERMUX__SE_PROCESS_CONTEXT=u:r:untrusted_app_27:s0:c127,c257,c512,c768
TERMUX__UID=10383
TERMUX__USER_ID=0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://mirrors.hust.edu.cn/termux/apt/termux-main stable main
# tur-repo (sources.list.d/tur.list)
deb https://tur.kcubeterm.com tur-packages tur tur-on-device tur-continuous
# x11-repo (sources.list.d/x11.list)
deb https://mirrors.hust.edu.cn/termux/apt/termux-x11 x11 main
Updatable packages:
All packages up to date
termux-tools version:
1.44.1
Android version:
14
Kernel build information:
Linux localhost 6.1.75-android14-11 #1 SMP PREEMPT Mon Aug 19 04:01:50 UTC 2024 aarch64 Android
Device manufacturer:
samsung
Device model:
SM-S926N
LD Variables:
LD_LIBRARY_PATH=
LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.so
Installed termux plugins:
com.termux.widget versionCode:13
com.termux.x11 versionCode:14
com.termux.api versionCode:51
```

**Comment by kyufie** [comment]
Try granting both permissions.

**Comment by hansm629** [comment]
@kyufie 
Unfortunately, I've enabled both permissions and it still doesn't work...

I'm not sure what I'm missing... 😭

**Comment by agnostic-apollo** [comment]
> ```
> mount | awk '/\/mnt\/media_rw/ && split($3, a, /\//){print "/storage/" a[4]}'
> /storage/7EEB-CB06
> $ ls /storage/7EEB-CB06
> "/storage/7EEB-CB06": No such file or directory
> ```

From the linked comment above.

>Unreliable/Removable volumes like USB OTG devices that are only available on the /mnt/media_rw paths with their own filesystem (vfat/exfat) are assigned the root (0) owner and external_storage (1077) group.

You are getting an `/mnt` path and replacing it with a  `/storage` path, which wouldn't exist as android never fuse mounts on it for usb drives like it does for sd cards.

**Comment by agnostic-apollo** [comment]
There is also a script given to find the usb drive path properly.

https://github.com/termux/termux-app/issues/71#issuecomment-2026147751

**Comment by hansm629** [comment]
@agnostic-apollo 
I added the following script to `~/.bashrc`, but I still don't have permission to access it.

Are there any parts of the script that need to be added or modified?

**<~/.bashrc>**
```shell
export LANG=ko_KR.UTF-8
export LC_MONETARY=ko_KR.UTF-8
export LC_PAPER=ko_KR.UTF-8
export LC_NAME=ko_KR.UTF-8
export LC_ADDRESS=ko_KR.UTF-8
export LC_TELEPHONE=ko_KR.UTF-8
export LC_MEASUREMENT=ko_KR.UTF-8
export LC_IDENTIFICATION=ko_KR.UTF-8
export LC_ALL=
export XDG_CONFIG_HOME=/data/data/com.termux/files/home/.config
export XMODIFIERS=@im=fcitx5
export GTK_IM_MODULE=fcitx5
export QT_IM_MODULE=fcitx5

find_most_recent_unreliable_storage_mount() {

    local return_value

    local i
    local storage_mounts_string
    local storage_mount
    local storage_mount_basename
    local unreliable_storage_mount
    local android_build_version_sdk
    local valid_number_regex='^[0-9]+$'

    local -a storage_mounts_array=()

    # Find android sdk/os version
    # Termux app exports ANDROID__BUILD_VERSION_SDK for `>= v0.119.0`
    if [[ ! "$ANDROID__BUILD_VERSION_SDK" =~ $valid_number_regex ]]; then
        android_build_version_sdk="$(unset LD_LIBRARY_PATH; unset LD_PRELOAD; getprop "ro.build.version.sdk")"
        return_value=$?
        if [ $return_value -ne 0 ] || [[ ! "$android_build_version_sdk" =~ $valid_number_regex ]]; then
            echo "Failure while finding \"ro.build.version.sdk\" property" 1>&2
            echo "ANDROID__BUILD_VERSION_SDK = \"$android_build_version_sdk\"" 1>&2
            if [ $return_value -eq 0 ]; then
                return_value=1
            fi
            return $return_value
        fi
    else
        android_build_version_sdk="$ANDROID__BUILD_VERSION_SDK"
    fi

    # If on Android `< 11`
    if [ "$android_build_version_sdk" -lt 30 ]; then
        echo "Cannot find unreliable storage mounts on Android sdk version '< 30'. Current sdk version is '$android_build_version_sdk'." 1>&2
        return 1
    fi

    storage_mounts_string="$(grep -E "^[^ ]+ /mnt/media_rw/[A-Z0-9]{4}-[A-Z0-9]{4} " /proc/mounts | cut -d' ' -f2)"
    return_value=$?
    if [ $return_value -ne 0 ]; then
        echo "Failure while finding most recent unreliable storage mount" 1>&2
        return $return_value
    fi

    if [ -z "$storage_mounts_string" ]; then
        echo "Failed to find any reliable or unreliable storage mounts" 1>&2
        return 1
    fi

    IFS=$'\n' read -r -d '' -a storage_mounts_array <<< "$storage_mounts_string"

    for (( i=${#storage_mounts_array[@]} - 1; i >= 0; i-- )) ; do
        storage_mount="${storage_mounts_array[i]}"
        storage_mount_basename="${storage_mount##*/}"

        grep -qE "^[^ ]+ /storage/$storage_mount_basename " /proc/mounts
        return_value=$?
        if [ $return_value -eq 1 ]; then
            unreliable_storage_mount="$storage_mount"
            break
        elif [ $return_value -ne 0 ]; then
            echo "Failure while finding if storage mount is unreliable" 1>&2
            return $return_value
        fi
    done

    if [ -z "$unreliable_storage_mount" ]; then
        echo "Failed to find any unreliable storage mounts" 1>&2
        return 1
    fi

    echo "$unreliable_storage_mount"

}

``` 




**<SD Card Permission denied>**
```
$ df | grep vold
/dev/block/vold/public:179,1 192219136 29112320 163106816  16% /mnt/media_rw/3232-6230

$ ls /mnt/media_rw/3232-6230
"/mnt/media_rw/3232-6230": Permission denied (os error 13)
``` 



**<Termux info>**
```
Termux Variables:
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP__APK_FILE=/data/app/~~uWpBhZ-vqdWJ__WFV9jslw==/com.termux-nm8zBMA-Cv4kSqJI8_UJbA==/base.apk
TERMUX_APP__APK_RELEASE=GITHUB
TERMUX_APP__APP_VERSION_CODE=1020
TERMUX_APP__APP_VERSION_NAME=0.119.0-beta.1
TERMUX_APP__DATA_DIR=/data/user/0/com.termux
TERMUX_APP__IS_DEBUGGABLE_BUILD=true
TERMUX_APP__IS_INSTALLED_ON_EXTERNAL_STORAGE=false
TERMUX_APP__PACKAGE_NAME=com.termux
TERMUX_APP__PID=20003
TERMUX_APP__TARGET_SDK=28
TERMUX_VERSION=0.119.0-beta.1
TERMUX__SE_FILE_CONTEXT=u:object_r:app_data_file:s0:c222,c257,c512,c768
TERMUX__SE_INFO=default:targetSdkVersion=28:complete
TERMUX__SE_PROCESS_CONTEXT=u:r:untrusted_app_27:s0:c222,c257,c512,c768
TERMUX__UID=10478
TERMUX__USER_ID=0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://mirror.sunred.org/termux/termux-main stable main
# tur-repo (sources.list.d/tur.list)
deb https://tur.kcubeterm.com tur-packages tur tur-on-device tur-continuous
# x11-repo (sources.list.d/x11.list)
deb https://packages-cf.termux.dev/apt/termux-x11/ x11 main
Updatable packages:
termux-am/stable 0.8.0-1 all [upgradable from: 0.8.0]
termux-tools version:
1.44.1
Android version:
14
Kernel build information:
Linux localhost 5.15.123-android13-8-29539737-abX916NKOU4BXHB #1 SMP PREEMPT Fri Aug 30 12:32:27 UTC 2024 aarch64 Android
Device manufacturer:
samsung
Device model:
SM-X916N
LD Variables:
LD_LIBRARY_PATH=
LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.so
Installed termux plugins:
com.termux.widget versionCode:13
com.termux.x11 versionCode:14
com.termux.api versionCode:51

```

File: packages/libusb/termux-usb-support.patch [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
++static int linux_termux_scan_devices(struct libusb_context *ctx)
++{
++	struct libusb_device *dev;
++	struct linux_device_priv *priv;
++	uint8_t busnum, devaddr;
++	unsigned long session_id;
++	char *envval, *envval_end;
++	int fd, r;
++
++	assert((envval = getenv(TERMUX_USB_ENV)));
```
Comment (landfillbaby) [inline packages/libusb/termux-usb-support.patch:129]: the assert and assignment should be separate, just in case someone tries to build it with -NDEBUG

**Comment by twaik** [comment]
This will require updating [docs](https://wiki.termux.com/wiki/Termux-usb), @agnostic-apollo can you help with this after it gets merged please?

**Comment by twaik** [comment]
@kyufie I do not think you need `TERMUX` guard for this. This patch will be applied only in termux.

**Comment by hansm629** [comment]
@twaik 
I succeeded! It works!

The external `USB storage` through `USB OTG` has been mounted on a `without root device`!

Both read and write operations are possible!

Although the volume doesn’t automatically appear in the `Thunar` file manager, I can access it in real-time by manually entering the path!

@kyufie @agnostic-apollo 
Thank you so much for your help!

Screenshot_2024-10-20_19-41-39

File: packages/libusb/termux-usb-support.patch (hunk 1 above)
Comment (kyufie) [reply packages/libusb/termux-usb-support.patch:129]: Right, quite an oversight that was.

**Comment by kyufie** [comment]
@twaik
That was a residue when the changes are still in my local libusb repo.
I'll remove that and do some more cleanup sometime later.

**Comment by hansm629** [comment]
@agnostic-apollo 
Hello sir,
Thanks to your help, I was able to successfully mount USB storage via OTG on my non-root device!
I really appreciate it.

However, I have one issue I need to ask about.

Mount/unmount USB storage via OTG works perfectly, but I can't access the `MicroSDXC card on the device` due to a permission issue.

`/mnt/media_rw/3232-6230` = MicroSDXC on the device
`/mnt/media_rw/3EFF-C796` = USB storage mount via USB OTG adapter

Screenshot_2024-10-20_23-17-59


```c
/dev/block/vold/public:179,1 192219136 29112576 163106560  16% /mnt/media_rw/3232-6230
/dev/block/vold/public:8,97   30013440  6582544  23430896  22% /mnt/media_rw/3EFF-C796

$ ls /mnt/media_rw/3232-6230
"/mnt/media_rw/3232-6230": Permission denied (os error 13)

$ ls /mnt/media_rw/3EFF-C796
.rwxrwx---  128 root  7 Oct  2021  autorun.inf*
drwxrwx---    - root 21 Jan  2022  boot/
.rwxrwx--- 414k root  7 Oct  2021  bootmgr*
.rwxrwx--- 1.5M root  7 Oct  2021  bootmgr.efi*
.rwxrwx--- 1.1G root 29 Jun 01:19  'COSTA RICA IN 4K@60fps VP9 HDR.webm'*
drwxrwx---    - root 21 Jan  2022  efi/
drwxrwx---    - root 18 Oct  2023  LOST.DIR/
.rwxrwx---  74k root  7 Oct  2021  setup.exe*
drwxrwx---    - root 21 Jan  2022  sources/
drwxrwx---    - root 21 Jan  2022  support/
drwxrwx---    - root 21 Jan  2022  'System Volume Information'/
drwxrwx---    - root  7 Dec  2023  포트폴리오/

``` 



Is there something I might be missing?
the method for accessing the MicroSDXC on the device different by any?

Below is the script entered in my `.bashrc`:
```c
find_most_recent_unreliable_storage_mount() {

    local return_value

    local i
    local storage_mounts_string
    local storage_mount
    local storage_mount_basename
    local unreliable_storage_mount
    local android_build_version_sdk
    local valid_number_regex='^[0-9]+$'

    local -a storage_mounts_array=()

    # Find android sdk/os version
    # Termux app exports ANDROID__BUILD_VERSION_SDK for `>= v0.119.0`
    if [[ ! "$ANDROID__BUILD_VERSION_SDK" =~ $valid_number_regex ]]; then
        android_build_version_sdk="$(unset LD_LIBRARY_PATH; unset LD_PRELOAD; getprop "ro.build.version.sdk")"
        return_value=$?
        if [ $return_value -ne 0 ] || [[ ! "$android_build_version_sdk" =~ $valid_number_regex ]]; then
            echo "Failure while finding \"ro.build.version.sdk\" property" 1>&2
            echo "ANDROID__BUILD_VERSION_SDK = \"$android_build_version_sdk\"" 1>&2
            if [ $return_value -eq 0 ]; then
                return_value=1
            fi
            return $return_value
        fi
    else
        android_build_version_sdk="$ANDROID__BUILD_VERSION_SDK"
    fi

    # If on Android `< 11`
    if [ "$android_build_version_sdk" -lt 30 ]; then
        echo "Cannot find unreliable storage mounts on Android sdk version '< 30'. Current sdk version is '$android_build_version_sdk'." 1>&2
        return 1
    fi

    storage_mounts_string="$(grep -E "^[^ ]+ /mnt/media_rw/[A-Z0-9]{4}-[A-Z0-9]{4} " /proc/mounts | cut -d' ' -f2)"
    return_value=$?
    if [ $return_value -ne 0 ]; then
        echo "Failure while finding most recent unreliable storage mount" 1>&2
        return $return_value
    fi

    if [ -z "$storage_mounts_string" ]; then
        echo "Failed to find any reliable or unreliable storage mounts" 1>&2
        return 1
    fi

    IFS=$'\n' read -r -d '' -a storage_mounts_array <<< "$storage_mounts_string"

    for (( i=${#storage_mounts_array[@]} - 1; i >= 0; i-- )) ; do
        storage_mount="${storage_mounts_array[i]}"
        storage_mount_basename="${storage_mount##*/}"

        grep -qE "^[^ ]+ /storage/$storage_mount_basename " /proc/mounts
        return_value=$?
        if [ $return_value -eq 1 ]; then
            unreliable_storage_mount="$storage_mount"
            break
        elif [ $return_value -ne 0 ]; then
            echo "Failure while finding if storage mount is unreliable" 1>&2
            return $return_value
        fi
    done

    if [ -z "$unreliable_storage_mount" ]; then
        echo "Failed to find any unreliable storage mounts" 1>&2
        return 1
    fi

    echo "$unreliable_storage_mount"

}
df | grep vold
```

**Comment by agnostic-apollo** [comment]
>This will require updating docs, @agnostic-apollo can you help with this after it gets merged please?

Send a MediaWiki formatted text and I'll upload.

**Comment by agnostic-apollo** [comment]
Welcome.

Read my post in the linked issue, external sd cards (reliable storages) are not accessible even with the permission, although some devices may allow access, like samsung. Only external hd (unreliable storages) will be accessible.

The `find_most_recent_unreliable_storage_mount()` is a shell function to get the most recent external hd mnt path, it does not give you special permissions, and adding it to `.bashrc` won't do anything alone, you need to call it by running `find_most_recent_unreliable_storage_mount` in a **`bash`** shell as a command and it will print the path so that you need to do df/vold shenanigans.

**Comment by hansm629** [comment]
@agnostic-apollo 
Hello sir,
Thank you for your kind and detailed comment.

After checking with the command below, it seems that the USB storage connected via USB OTG is recognized, but the MicroSDXC card installed on the device is not recognized.

If it's displayed as shown below, should I assume that even with permissions, it's impossible to access the MicroSDXC card directly installed on the device (as opposed to via OTG)?


```c
/dev/block/vold/public:179,1 192219136 29112576 163106560  16% /mnt/media_rw/3232-6230

$ find_most_recent_unreliable_storage_mount
Failed to find any unreliable storage mounts

```

**Comment by kyufie** [comment]
@hansm629 
I'm curious, can you replace `/mnt/media_rw` for your microSD card with `/storage` and see if it works?

**Comment by hansm629** [comment]
@kyufie 
Thank you, sir!
Thanks to your help, I was able to resolve the issue!

By converting `/mnt/media_rw/` to `/storage/`, I can now access it, and both reading and writing on the `MicroSDXC` card work perfectly!
I also pinned it as a bookmark in the `Thunar` file manager, so now I can easily access it anytime!
Screenshot_2024-10-21_11-50-14

**Comment by kyufie** [comment]
That should be it.
Is it ready for merging?

**Review by TomJo2000 — APPROVED** [review]
This should be good to merge.
I'd like to give it another day for final comments/reviews.

**Review by landfillbaby — APPROVED** [review]

**Comment by TomJo2000** [comment]
Ah crap it looks like we have a rebase conflict.
I just tried to rebase the branch on top of the latest state from master and it's throwing up an error.
Let me see if I can find and resolve the issue.

**Comment by TomJo2000** [comment]
Thank you for your contribution @kyufie.

<h1><em></em></h1> 

To avoid last minute rebasing errors in the future I would suggest periodically resyncing your development fork with the upstream `master` branch of termux/termux-packages.

This is how I do that for my fork:
```yml
# One time setup to add the upstream remote for fetching new commits.
git remote add upstream https://github.com/termux/termux-packages.git

# Then to resync the fork:
# On your fork's master branch
git fetch --all -v && git rebase upstream/master
# and to push that back to your development fork on GitHub:
git push

# You can then switch back to your dev branch and rebase it on top of local master,
# which is now up to date, and force push it to update your PR
git switch branch-name
git rebase master
git push --force-with-lease=branch-name 
# --force would also work, but force with lease is preferred to not accidentally clobber anything
```
You can probably set up git aliases or hooks to make this more convenient.

**Comment by kyufie** [comment]
Thanks, I'll keep that in mind.

**Comment by landfillbaby** [comment]
to make that happen by default, here's an excerpt from my `~/.gitconfig`:

```
[branch]
	autosetupmerge = always
	autosetuprebase = always
[pull]
	rebase = true
[rebase]
	autostash = true
```

then `git pull` just rebases automatically

**Comment by TomJo2000** [comment]
Did not know about branch `autosetupmerge` and `autosetuprebase`.
Thanks for the tip.

**Comment by wildxxxx** [comment]
@kyufie

hello! I wanted to make a usb pass-through in qemu on termux, but the output of "info usbhost" says that it uses "port 0" and "Speed ? Mb/s". Which is why I can't use it in qemu. What should I do about it? Thank you.

**Pull request merged**